File: app/src/main/java/com/angogasapps/myfamily/objects/ChatVoicePlayer.java
package com.angogasapps.myfamily.objects;

import android.content.Context;
import android.media.MediaPlayer;

import com.angogasapps.myfamily.R;
import com.angogasapps.myfamily.firebase.ChatFunks;
import com.angogasapps.myfamily.firebase.interfaces.IOnEndCommunicationWithFirebase;


import java.io.File;

import es.dmoral.toasty.Toasty;


public class ChatVoicePlayer {
    private MediaPlayer mMediaPlayer;
    private Context context;
    private File voiceFile;

    public ChatVoicePlayer(Context context) {
        this.context = context;
    }

    public void play(String messageKey, String fileUrl, IOnEndPlay iOnEndPlay){
        voiceFile = new File(context.getFilesDir().getAbsolutePath(), messageKey);

        if (voiceFile.exists() && voiceFile.isFile() && voiceFile.length() > 0){
            startPlayer(iOnEndPlay);
        }else {
            try {
                voiceFile.createNewFile();
//                ChatFunks.getFileFromStorage(voiceFile, fileUrl, new IOnEndCommunicationWithFirebase() {
                ChatFunks.getVoiceFileFromStorage(voiceFile, messageKey, new IOnEndCommunicationWithFirebase() {
                    @Override
                    public void onSuccess() {
                        startPlayer(iOnEndPlay);
                    }

                    @Override
                    public void onFailure() {}
                });
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
    }

    private void startPlayer(IOnEndPlay iOnEndPlay){
        try {

            mMediaPlayer.setDataSource(voiceFile.getAbsolutePath());
            mMediaPlayer.prepare();
            mMediaPlayer.start();
            mMediaPlayer.setOnCompletionListener(mp -> {
                stop(iOnEndPlay);
            });

        } catch (Exception e) {
            e.printStackTrace();
            Toasty.error(context, R.string.something_went_wrong).show();
        }

    }

    public void stop(IOnEndPlay iOnEndPlay){
        try {
            mMediaPlayer.stop();
            mMediaPlayer.reset();
        }catch (Exception e){
            e.printStackTrace();
            Toasty.error(context, R.string.something_went_wrong).show();
        }
        iOnEndPlay.onEndPlay();
    }

    public void release(){
        mMediaPlayer.release();
    }

    public void init(){
        mMediaPlayer = new MediaPlayer();
    }


    public interface IOnEndPlay{
        void onEndPlay();
    }
}

File: app/src/main/java/com/angogasapps/myfamily/objects/ChatVoicePlayer.kt
package com.angogasapps.myfamily.objects

import android.content.Context
import android.media.MediaDataSource
import android.media.MediaPlayer
import android.media.MediaRecorder
import es.dmoral.toasty.Toasty
import com.angogasapps.myfamily.R
import com.angogasapps.myfamily.app.appComponent
import com.angogasapps.myfamily.network.interfaces.chat.ChatVoiceGetter
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import java.io.File
import java.lang.Exception
import javax.inject.Inject

class ChatVoicePlayer(private val context: Context) {
    private lateinit var mMediaPlayer: MediaPlayer
    private lateinit var voiceFile: File
    @Inject
    lateinit var chatVoiceGetter: ChatVoiceGetter

    init {
        appComponent.inject(this)
    }


    fun play(messageKey: String, fileUrl: String, onEndPlay: () -> Unit) {
        voiceFile = File(context.filesDir.absolutePath, messageKey)
        if (voiceFile.exists() && voiceFile.isFile && voiceFile.length() > 0) {
            startPlayer(onEndPlay)
        } else {
            try {
                voiceFile.createNewFile()
                chatVoiceGetter.getVoiceFileFromStorage(
                    voiceFile,
                    messageKey,
                    onSuccess = {
                        startPlayer(onEndPlay)
                    }
                )


            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
    }

    private fun startPlayer(onEndPlay: () -> Unit) {
        try {
            mMediaPlayer.setDataSource(voiceFile.absolutePath /*+ ".mp3"*/)


            mMediaPlayer.prepare()
            mMediaPlayer.start()
            mMediaPlayer.setOnCompletionListener { mp: MediaPlayer? -> stop(onEndPlay) }
        } catch (e: Exception) {
            e.printStackTrace()
            GlobalScope.launch(Dispatchers.Main) {
                Toasty.error(context, R.string.something_went_wrong).show()
            }
        }
    }

    fun stop(onEndPlay: () -> Unit) {
        try {
            mMediaPlayer.stop()
            mMediaPlayer.reset()
        } catch (e: Exception) {
            e.printStackTrace()
            Toasty.error(context, R.string.something_went_wrong).show()
        }
        onEndPlay()
    }

    fun release() {
        mMediaPlayer.release()
    }

    fun init() {
        mMediaPlayer = MediaPlayer()
    }
}